File: rss.py
# rss.py
# RSS Reader (Фильтрация новостей) на Python

import sys
import os
import json
import time
import datetime
import argparse
import xml.etree.ElementTree as ET
import urllib.request
import subprocess
import re
from typing import List, Dict, Optional, Any

# ANSI-цвета (тёмная тема)
RESET = "\033[0m"
BOLD = "\033[1m"
RED = "\033[91m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
BLUE = "\033[94m"
MAGENTA = "\033[95m"
CYAN = "\033[96m"
WHITE = "\033[97m"
GRAY = "\033[90m"

def colorize(text: str, color: str) -> str:
    return f"{color}{text}{RESET}"

class Filter:
    def __init__(self, keywords=None, days=None, source=None, unread_only=False, read_only=False):
        self.keywords = keywords or []
        self.days = days
        self.source = source
        self.unread_only = unread_only
        self.read_only = read_only

    def to_dict(self):
        return {
            "keywords": self.keywords,
            "days": self.days,
            "source": self.source,
            "unread_only": self.unread_only,
            "read_only": self.read_only
        }

    @classmethod
    def from_dict(cls, data):
        return cls(
            keywords=data.get("keywords", []),
            days=data.get("days"),
            source=data.get("source"),
            unread_only=data.get("unread_only", False),
            read_only=data.get("read_only", False)
        )

class RssReader:
    def __init__(self, data_file="rss_data.json"):
        self.data_file = data_file
        self.data = self._load()
        self.next_feed_id = self.data.get("next_feed_id", 1)
        self.next_item_id = self.data.get("next_item_id", 1)
        self.filters = Filter.from_dict(self.data.get("filters", {}))

    def _load(self):
        if os.path.exists(self.data_file):
            with open(self.data_file, 'r', encoding='utf-8') as f:
                return json.load(f)
        return {"feeds": [], "next_feed_id": 1, "next_item_id": 1, "filters": {}}

    def _save(self):
        self.data["next_feed_id"] = self.next_feed_id
        self.data["next_item_id"] = self.next_item_id
        self.data["filters"] = self.filters.to_dict()
        with open(self.data_file, 'w', encoding='utf-8') as f:
            json.dump(self.data, f, indent=2, ensure_ascii=False)

    def _fetch_feed(self, url):
        try:
            with urllib.request.urlopen(url, timeout=10) as response:
                content = response.read().decode('utf-8')
                return content
        except Exception as e:
            raise Exception(f"Не удалось загрузить ленту: {e}")

    def _parse_feed(self, content):
        root = ET.fromstring(content)
        if root.tag == 'rss':
            return self._parse_rss(root)
        elif root.tag.endswith('feed'):
            return self._parse_atom(root)
        else:
            raise Exception("Неизвестный формат ленты")

    def _parse_rss(self, root):
        channel = root.find('channel')
        title = channel.find('title').text if channel.find('title') is not None else "Без названия"
        items = []
        for item in channel.findall('item'):
            title_elem = item.find('title')
            link_elem = item.find('link')
            pub_date_elem = item.find('pubDate')
            desc_elem = item.find('description')
            categories = [cat.text for cat in item.findall('category') if cat.text]
            items.append({
                'title': title_elem.text if title_elem is not None else "Без заголовка",
                'link': link_elem.text if link_elem is not None else "",
                'pubDate': pub_date_elem.text if pub_date_elem is not None else "",
                'description': desc_elem.text if desc_elem is not None else "",
                'categories': categories
            })
        return title, items

    def _parse_atom(self, root):
        title = root.find('title').text if root.find('title') is not None else "Без названия"
        items = []
        for entry in root.findall('entry'):
            title_elem = entry.find('title')
            link_elem = entry.find('link')
            pub_date_elem = entry.find('published')
            if pub_date_elem is None:
                pub_date_elem = entry.find('updated')
            desc_elem = entry.find('summary')
            if desc_elem is None:
                desc_elem = entry.find('content')
            categories = [cat.get('term') for cat in entry.findall('category') if cat.get('term')]
            items.append({
                'title': title_elem.text if title_elem is not None else "Без заголовка",
                'link': link_elem.get('href') if link_elem is not None else "",
                'pubDate': pub_date_elem.text if pub_date_elem is not None else "",
                'description': desc_elem.text if desc_elem is not None else "",
                'categories': categories
            })
        return title, items

    def add_feed(self, url, title=None):
        if not title:
            try:
                content = self._fetch_feed(url)
                title, _ = self._parse_feed(content)
            except:
                title = url
        feed = {
            "id": self.next_feed_id,
            "title": title,
            "url": url,
            "items": [],
            "last_fetch": None
        }
        self.data["feeds"].append(feed)
        self.next_feed_id += 1
        self._save()
        return feed

    def remove_feed(self, feed_id=None, url=None):
        for i, f in enumerate(self.data["feeds"]):
            if (feed_id is not None and f["id"] == feed_id) or (url is not None and f["url"] == url):
                del self.data["feeds"][i]
                self._save()
                return True
        return False

    def list_feeds(self):
        return self.data["feeds"]

    def fetch_all(self, filter_obj: Optional[Filter] = None):
        if filter_obj is None:
            filter_obj = self.filters
        for feed in self.data["feeds"]:
            try:
                content = self._fetch_feed(feed["url"])
                title, items = self._parse_feed(content)
                if title != feed["title"]:
                    feed["title"] = title
                new_items = []
                for item in items:
                    exists = any(i["link"] == item["link"] for i in feed["items"])
                    if not exists:
                        item["id"] = self.next_item_id
                        item["read"] = False
                        new_items.append(item)
                        self.next_item_id += 1
                feed["items"].extend(new_items)
                feed["last_fetch"] = datetime.datetime.now().isoformat()
            except Exception as e:
                print(colorize(f"Ошибка при загрузке {feed['url']}: {e}", RED))
        self._save()
        # Применяем фильтры
        all_items = []
        for feed in self.data["feeds"]:
            for item in feed["items"]:
                all_items.append((feed["title"], item))
        # Фильтрация
        all_items = self._apply_filters(all_items, filter_obj)
        # Сортировка по дате (новые сверху)
        all_items.sort(key=lambda x: x[1].get("pubDate", ""), reverse=True)
        return all_items

    def _apply_filters(self, items, filter_obj: Filter):
        result = []
        for feed_title, item in items:
            # Фильтр по ключевым словам
            if filter_obj.keywords:
                text = (item.get("title", "") + " " + item.get("description", "") + " " + " ".join(item.get("categories", []))).lower()
                if not any(kw.lower() in text for kw in filter_obj.keywords):
                    continue
            # Фильтр по дате
            if filter_obj.days:
                try:
                    pub = datetime.datetime.fromisoformat(item.get("pubDate", "").replace('Z', '+00:00'))
                    if (datetime.datetime.now(datetime.timezone.utc) - pub).days > filter_obj.days:
                        continue
                except:
                    pass
            # Фильтр по источнику
            if filter_obj.source and filter_obj.source.lower() not in feed_title.lower():
                continue
            # Фильтр по статусу
            if filter_obj.unread_only and item.get("read", False):
                continue
            if filter_obj.read_only and not item.get("read", False):
                continue
            result.append((feed_title, item))
        return result

    def get_item(self, item_id):
        for feed in self.data["feeds"]:
            for item in feed["items"]:
                if item["id"] == item_id:
                    return feed["title"], item
        return None, None

    def mark_read(self, item_id):
        for feed in self.data["feeds"]:
            for item in feed["items"]:
                if item["id"] == item_id:
                    item["read"] = True
                    self._save()
                    return True
        return False

    def open_link(self, url):
        if sys.platform == 'darwin':
            subprocess.run(['open', url])
        elif sys.platform == 'win32':
            subprocess.run(['start', url], shell=True)
        else:
            subprocess.run(['xdg-open', url])

    def export_opml(self, filename="subscriptions.opml"):
        opml = '<?xml version="1.0" encoding="UTF-8"?>\n'
        opml += '<opml version="1.0">\n<head>\n<title>RSS Subscriptions</title>\n</head>\n<body>\n'
        for feed in self.data["feeds"]:
            opml += f'<outline text="{feed["title"]}" title="{feed["title"]}" type="rss" xmlUrl="{feed["url"]}"/>\n'
        opml += '</body>\n</opml>'
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(opml)
        return filename

    def import_opml(self, filename):
        try:
            tree = ET.parse(filename)
            root = tree.getroot()
            for outline in root.findall('.//outline'):
                url = outline.get('xmlUrl')
                title = outline.get('text') or outline.get('title') or url
                if url:
                    self.add_feed(url, title)
            self._save()
            return True
        except Exception as e:
            print(colorize(f"Ошибка импорта OPML: {e}", RED))
            return False

    def save_filters(self):
        self._save()
        print(colorize("Фильтры сохранены", GREEN))

    def load_filters(self):
        # уже загружены в __init__
        print(colorize("Фильтры загружены", GREEN))

    def clear_filters(self):
        self.filters = Filter()
        self._save()
        print(colorize("Фильтры очищены", GREEN))

def main():
    parser = argparse.ArgumentParser(description="RSS Reader (Фильтрация новостей)")
    parser.add_argument("command", choices=["add", "remove", "list", "fetch", "read", "filter", "save-filters", "load-filters", "export", "import", "help"],
                        help="Команда")
    parser.add_argument("--url", help="URL ленты")
    parser.add_argument("--title", help="Название ленты")
    parser.add_argument("--id", type=int, help="ID ленты или новости")
    parser.add_argument("--file", help="Имя файла для импорта/экспорта")
    parser.add_argument("--data", default="rss_data.json", help="Файл данных")
    parser.add_argument("--open", action="store_true", default=True, help="Открыть ссылку в браузере")
    parser.add_argument("--text", action="store_true", help="Показать текст в консоли")
    # Опции фильтрации
    parser.add_argument("--keyword", action="append", help="Ключевое слово для фильтрации (можно несколько)")
    parser.add_argument("--days", type=int, help="Показывать новости за последние N дней")
    parser.add_argument("--source", help="Фильтр по источнику (название ленты)")
    parser.add_argument("--unread", action="store_true", help="Показывать только непрочитанные")
    parser.add_argument("--read", action="store_true", help="Показывать только прочитанные")
    parser.add_argument("--clear-filters", action="store_true", help="Очистить все фильтры")
    args = parser.parse_args()

    reader = RssReader(args.data)

    if args.command == "help":
        print(__doc__)
        sys.exit(0)

    # Обработка фильтров
    if args.clear_filters:
        reader.clear_filters()
        return

    if args.command == "filter" or args.keyword or args.days or args.source or args.unread or args.read:
        # Обновляем фильтры
        if args.keyword:
            reader.filters.keywords = args.keyword
        if args.days:
            reader.filters.days = args.days
        if args.source:
            reader.filters.source = args.source
        if args.unread:
            reader.filters.unread_only = True
            reader.filters.read_only = False
        if args.read:
            reader.filters.read_only = True
            reader.filters.unread_only = False
        reader._save()
        print(colorize("Фильтры обновлены", GREEN))
        return

    if args.command == "save-filters":
        reader.save_filters()
        return

    if args.command == "load-filters":
        reader.load_filters()
        return

    if args.command == "add":
        if not args.url:
            print("Требуется --url")
            sys.exit(1)
        feed = reader.add_feed(args.url, args.title)
        print(colorize(f"Лента добавлена: {feed['title']} (ID {feed['id']})", GREEN))

    elif args.command == "remove":
        if args.id is None and not args.url:
            print("Требуется --id или --url")
            sys.exit(1)
        if reader.remove_feed(args.id, args.url):
            print(colorize("Лента удалена", GREEN))
        else:
            print(colorize("Лента не найдена", RED))

    elif args.command == "list":
        feeds = reader.list_feeds()
        if not feeds:
            print("Нет лент.")
        else:
            print(colorize("Список лент:", BOLD + CYAN))
            for f in feeds:
                print(f"  {colorize(str(f['id']), GREEN)}: {f['title']} ({f['url']})")

    elif args.command == "fetch":
        filter_obj = reader.filters  # используем текущие
        items = reader.fetch_all(filter_obj)
        if not items:
            print("Новостей нет (возможно, фильтры слишком строгие).")
        else:
            print(colorize("Новости (отфильтрованные):", BOLD + CYAN))
            for feed_title, item in items:
                status = "🔵" if not item.get("read", False) else "⚪"
                date = item.get("pubDate", "")[:16]
                print(f"{status} {colorize(str(item['id']), GREEN)} | {colorize(item['title'][:60], WHITE)} | {colorize(feed_title, GRAY)} | {colorize(date, YELLOW)}")

    elif args.command == "read":
        if args.id is None:
            print("Требуется --id")
            sys.exit(1)
        feed_title, item = reader.get_item(args.id)
        if not item:
            print(colorize("Новость не найдена", RED))
        else:
            reader.mark_read(args.id)
            print(colorize(f"Заголовок: {item['title']}", BOLD + WHITE))
            print(colorize(f"Дата: {item.get('pubDate', '')}", YELLOW))
            print(colorize(f"Источник: {feed_title}", GRAY))
            if args.text:
                print(colorize("Содержание:", BOLD))
                print(item.get("description", "Нет описания"))
            else:
                print(colorize(f"Ссылка: {item.get('link', '')}", CYAN))
                if item.get('link'):
                    try:
                        reader.open_link(item['link'])
                        print(colorize("Ссылка открыта в браузере", GREEN))
                    except:
                        print(colorize("Не удалось открыть браузер", RED))

    elif args.command == "export":
        filename = args.file or "subscriptions.opml"
        reader.export_opml(filename)
        print(colorize(f"Подписки экспортированы в {filename}", GREEN))

    elif args.command == "import":
        if not args.file:
            print("Требуется --file")
            sys.exit(1)
        if reader.import_opml(args.file):
            print(colorize("Подписки импортированы", GREEN))

if __name__ == "__main__":
    main()

File: rss.js
// rss.js
// RSS Reader (Фильтрация новостей) на JavaScript (Node.js)

const fs = require('fs');
const https = require('https');
const http = require('http');
const { URL } = require('url');
const { exec } = require('child_process');
const readline = require('readline');

// ANSI-цвета (тёмная тема)
const RESET = '\x1b[0m';
const BOLD = '\x1b[1m';
const RED = '\x1b[91m';
const GREEN = '\x1b[92m';
const YELLOW = '\x1b[93m';
const BLUE = '\x1b[94m';
const MAGENTA = '\x1b[95m';
const CYAN = '\x1b[96m';
const WHITE = '\x1b[97m';
const GRAY = '\x1b[90m';

function colorize(text, color) {
    return `${color}${text}${RESET}`;
}

class Filter {
    constructor(keywords = [], days = null, source = null, unreadOnly = false, readOnly = false) {
        this.keywords = keywords;
        this.days = days;
        this.source = source;
        this.unreadOnly = unreadOnly;
        this.readOnly = readOnly;
    }
}

class RssReader {
    constructor(dataFile = 'rss_data.json') {
        this.dataFile = dataFile;
        this.data = this.load();
        this.nextFeedId = this.data.next_feed_id || 1;
        this.nextItemId = this.data.next_item_id || 1;
        this.filters = new Filter(
            this.data.filters?.keywords || [],
            this.data.filters?.days || null,
            this.data.filters?.source || null,
            this.data.filters?.unread_only || false,
            this.data.filters?.read_only || false
        );
    }

    load() {
        try {
            if (fs.existsSync(this.dataFile)) {
                return JSON.parse(fs.readFileSync(this.dataFile, 'utf-8'));
            }
        } catch (e) {}
        return { feeds: [], next_feed_id: 1, next_item_id: 1, filters: {} };
    }

    save() {
        this.data.next_feed_id = this.nextFeedId;
        this.data.next_item_id = this.nextItemId;
        this.data.filters = {
            keywords: this.filters.keywords,
            days: this.filters.days,
            source: this.filters.source,
            unread_only: this.filters.unreadOnly,
            read_only: this.filters.readOnly
        };
        fs.writeFileSync(this.dataFile, JSON.stringify(this.data, null, 2), 'utf-8');
    }

    fetchFeed(url) {
        return new Promise((resolve, reject) => {
            const parsed = new URL(url);
            const client = parsed.protocol === 'https:' ? https : http;
            client.get(url, (res) => {
                let data = '';
                res.on('data', chunk => data += chunk);
                res.on('end', () => resolve(data));
            }).on('error', reject);
        });
    }

    parseFeed(content) {
        // Простой парсинг через регулярки (без библиотек)
        const items = [];
        let title = 'Без названия';
        const titleMatch = content.match(/<title>(.*?)<\/title>/i);
        if (titleMatch) title = titleMatch[1];
        const itemRegex = /<(item|entry)[^>]*>([\s\S]*?)<\/(item|entry)>/gi;
        let match;
        while ((match = itemRegex.exec(content)) !== null) {
            const itemContent = match[2];
            const item = {};
            const titleEl = itemContent.match(/<title>(.*?)<\/title>/i);
            const linkEl = itemContent.match(/<link[^>]*>(.*?)<\/link>/i) || itemContent.match(/<link[^>]*href="([^"]*)"/i);
            const pubEl = itemContent.match(/<(pubDate|published|updated)>(.*?)<\/(pubDate|published|updated)>/i);
            const descEl = itemContent.match(/<(description|summary|content)>(.*?)<\/(description|summary|content)>/i);
            const catEls = itemContent.match(/<category[^>]*>([^<]*)<\/category>/gi);
            item.title = titleEl ? titleEl[1].trim() : 'Без заголовка';
            item.link = linkEl ? (linkEl[1] || linkEl[0]) : '';
            item.pubDate = pubEl ? pubEl[2].trim() : '';
            item.description = descEl ? descEl[2].trim() : '';
            item.categories = catEls ? catEls.map(c => c.replace(/<[^>]*>/g,'').trim()) : [];
            items.push(item);
        }
        return { title, items };
    }

    addFeed(url, title) {
        if (!title) {
            return this.fetchFeed(url).then(content => {
                const parsed = this.parseFeed(content);
                const feedTitle = parsed.title || url;
                return this._addFeed(url, feedTitle);
            }).catch(() => {
                return this._addFeed(url, url);
            });
        } else {
            return Promise.resolve(this._addFeed(url, title));
        }
    }

    _addFeed(url, title) {
        const feed = {
            id: this.nextFeedId++,
            title: title,
            url: url,
            items: [],
            last_fetch: null
        };
        this.data.feeds.push(feed);
        this.save();
        return feed;
    }

    removeFeed(feedId, url) {
        const idx = this.data.feeds.findIndex(f => (feedId && f.id === feedId) || (url && f.url === url));
        if (idx !== -1) {
            this.data.feeds.splice(idx, 1);
            this.save();
            return true;
        }
        return false;
    }

    listFeeds() {
        return this.data.feeds;
    }

    fetchAll(filterObj = null) {
        if (!filterObj) filterObj = this.filters;
        const promises = this.data.feeds.map(feed => {
            return this.fetchFeed(feed.url).then(content => {
                const parsed = this.parseFeed(content);
                if (parsed.title !== feed.title) feed.title = parsed.title;
                const newItems = [];
                for (const item of parsed.items) {
                    const exists = feed.items.some(i => i.link === item.link);
                    if (!exists) {
                        item.id = this.nextItemId++;
                        item.read = false;
                        newItems.push(item);
                    }
                }
                feed.items.push(...newItems);
                feed.last_fetch = new Date().toISOString();
            }).catch(err => {
                console.log(colorize(`Ошибка при загрузке ${feed.url}: ${err.message}`, RED));
            });
        });
        return Promise.all(promises).then(() => {
            this.save();
            const all = [];
            for (const feed of this.data.feeds) {
                for (const item of feed.items) {
                    all.push({ feed: feed.title, item });
                }
            }
            // Применяем фильтры
            const filtered = this._applyFilters(all, filterObj);
            filtered.sort((a,b) => (a.item.pubDate < b.item.pubDate ? 1 : -1));
            return filtered;
        });
    }

    _applyFilters(items, filterObj) {
        return items.filter(({ feed, item }) => {
            // Ключевые слова
            if (filterObj.keywords && filterObj.keywords.length > 0) {
                const text = (item.title + ' ' + item.description + ' ' + (item.categories || []).join(' ')).toLowerCase();
                if (!filterObj.keywords.some(kw => text.includes(kw.toLowerCase()))) return false;
            }
            // Дата
            if (filterObj.days) {
                try {
                    const pub = new Date(item.pubDate);
                    const now = new Date();
                    const diffDays = (now - pub) / (1000 * 60 * 60 * 24);
                    if (diffDays > filterObj.days) return false;
                } catch (e) {}
            }
            // Источник
            if (filterObj.source && !feed.toLowerCase().includes(filterObj.source.toLowerCase())) return false;
            // Статус
            if (filterObj.unreadOnly && item.read) return false;
            if (filterObj.readOnly && !item.read) return false;
            return true;
        });
    }

    getItem(itemId) {
        for (const feed of this.data.feeds) {
            for (const item of feed.items) {
                if (item.id === itemId) {
                    return { feedTitle: feed.title, item };
                }
            }
        }
        return null;
    }

    markRead(itemId) {
        const res = this.getItem(itemId);
        if (res) {
            res.item.read = true;
            this.save();
            return true;
        }
        return false;
    }

    openLink(url) {
        const cmd = process.platform === 'darwin' ? 'open' :
                    process.platform === 'win32' ? 'start' : 'xdg-open';
        exec(`${cmd} "${url}"`);
    }

    exportOpml(filename = 'subscriptions.opml') {
        let opml = '<?xml version="1.0" encoding="UTF-8"?>\n';
        opml += '<opml version="1.0">\n<head>\n<title>RSS Subscriptions</title>\n</head>\n<body>\n';
        for (const feed of this.data.feeds) {
            opml += `<outline text="${feed.title}" title="${feed.title}" type="rss" xmlUrl="${feed.url}"/>\n`;
        }
        opml += '</body>\n</opml>';
        fs.writeFileSync(filename, opml, 'utf-8');
        return filename;
    }

    importOpml(filename) {
        const content = fs.readFileSync(filename, 'utf-8');
        const regex = /<outline[^>]*xmlUrl="([^"]*)"[^>]*text="([^"]*)"[^>]*\/?>/gi;
        let match;
        let count = 0;
        while ((match = regex.exec(content)) !== null) {
            const url = match[1];
            const title = match[2] || url;
            this._addFeed(url, title);
            count++;
        }
        this.save();
        return count;
    }

    saveFilters() {
        this.save();
        console.log(colorize('Фильтры сохранены', GREEN));
    }

    loadFilters() {
        // уже загружены в конструкторе
        console.log(colorize('Фильтры загружены', GREEN));
    }

    clearFilters() {
        this.filters = new Filter();
        this.save();
        console.log(colorize('Фильтры очищены', GREEN));
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length === 0 || args[0] === 'help') {
        console.log(`Использование: node rss.js <команда> [опции]
  add       --url <url> [--title <title>]
  remove    --id <id> | --url <url>
  list
  fetch     [--keyword <kw>] [--days N] [--source <src>] [--unread] [--read]
  read      --id <id> [--text]
  filter    [--keyword <kw>] [--days N] [--source <src>] [--unread] [--read]
  save-filters
  load-filters
  export    [--file <file>]
  import    --file <file>
  help`);
        process.exit(0);
    }

    const command = args[0];
    const options = {};
    for (let i = 1; i < args.length; i++) {
        const arg = args[i];
        if (arg.startsWith('--')) {
            const key = arg.slice(2);
            const value = args[++i];
            options[key] = value;
        }
    }

    const dataFile = options.data || 'rss_data.json';
    const reader = new RssReader(dataFile);

    // Обработка фильтров
    if (command === 'filter' || options.keyword || options.days || options.source || options.unread || options.read) {
        if (options.keyword) {
            const kw = Array.isArray(options.keyword) ? options.keyword : [options.keyword];
            reader.filters.keywords = kw;
        }
        if (options.days) reader.filters.days = parseInt(options.days);
        if (options.source) reader.filters.source = options.source;
        if (options.unread) { reader.filters.unreadOnly = true; reader.filters.readOnly = false; }
        if (options.read) { reader.filters.readOnly = true; reader.filters.unreadOnly = false; }
        reader.save();
        console.log(colorize('Фильтры обновлены', GREEN));
        return;
    }

    if (command === 'save-filters') {
        reader.saveFilters();
        return;
    }
    if (command === 'load-filters') {
        reader.loadFilters();
        return;
    }

    switch (command) {
        case 'add': {
            if (!options.url) { console.error('Требуется --url'); process.exit(1); }
            reader.addFeed(options.url, options.title).then(feed => {
                console.log(colorize(`Лента добавлена: ${feed.title} (ID ${feed.id})`, GREEN));
            }).catch(err => console.error(colorize(`Ошибка: ${err.message}`, RED)));
            break;
        }
        case 'remove': {
            if (!options.id && !options.url) { console.error('Требуется --id или --url'); process.exit(1); }
            const id = options.id ? parseInt(options.id) : null;
            if (reader.removeFeed(id, options.url)) {
                console.log(colorize('Лента удалена', GREEN));
            } else {
                console.log(colorize('Лента не найдена', RED));
            }
            break;
        }
        case 'list': {
            const feeds = reader.listFeeds();
            if (!feeds.length) console.log('Нет лент.');
            else {
                console.log(colorize('Список лент:', BOLD + CYAN));
                for (const f of feeds) {
                    console.log(`  ${colorize(f.id, GREEN)}: ${f.title} (${f.url})`);
                }
            }
            break;
        }
        case 'fetch': {
            // Применяем текущие фильтры, но можно переопределить через опции
            const filterObj = reader.filters;
            reader.fetchAll(filterObj).then(items => {
                if (!items.length) {
                    console.log('Новостей нет (возможно, фильтры слишком строгие).');
                } else {
                    console.log(colorize('Новости (отфильтрованные):', BOLD + CYAN));
                    for (const { feed, item } of items) {
                        const status = item.read ? '⚪' : '🔵';
                        const date = (item.pubDate || '').slice(0,16);
                        console.log(`${status} ${colorize(item.id, GREEN)} | ${colorize(item.title.slice(0,60), WHITE)} | ${colorize(feed, GRAY)} | ${colorize(date, YELLOW)}`);
                    }
                }
            });
            break;
        }
        case 'read': {
            if (!options.id) { console.error('Требуется --id'); process.exit(1); }
            const id = parseInt(options.id);
            const res = reader.getItem(id);
            if (!res) {
                console.log(colorize('Новость не найдена', RED));
            } else {
                reader.markRead(id);
                console.log(colorize(`Заголовок: ${res.item.title}`, BOLD + WHITE));
                console.log(colorize(`Дата: ${res.item.pubDate}`, YELLOW));
                console.log(colorize(`Источник: ${res.feedTitle}`, GRAY));
                if (options.text) {
                    console.log(colorize('Содержание:', BOLD));
                    console.log(res.item.description || 'Нет описания');
                } else {
                    console.log(colorize(`Ссылка: ${res.item.link}`, CYAN));
                    if (res.item.link) {
                        try {
                            reader.openLink(res.item.link);
                            console.log(colorize('Ссылка открыта в браузере', GREEN));
                        } catch (e) {
                            console.log(colorize('Не удалось открыть браузер', RED));
                        }
                    }
                }
            }
            break;
        }
        case 'export': {
            const filename = options.file || 'subscriptions.opml';
            reader.exportOpml(filename);
            console.log(colorize(`Подписки экспортированы в ${filename}`, GREEN));
            break;
        }
        case 'import': {
            if (!options.file) { console.error('Требуется --file'); process.exit(1); }
            const count = reader.importOpml(options.file);
            console.log(colorize(`Импортировано ${count} подписок`, GREEN));
            break;
        }
        default:
            console.log('Неизвестная команда.');
    }
}

main();
